### PR TITLE
Compute expected ancestry proof size

### PR DESCRIPTION
**Pull request by serban300**
Related to: https://github.com/paritytech/polkadot-sdk/issues/4523

This is a follow-up to https://github.com/paritytech/polkadot-sdk/pull/5188

The main purpose of this PR is to add logic for computing the expected ancestry proof size . 

This PR also aligns some functions from the `ancestry_proof.rs` file with the `mmr.rs` file, which should make it easier to upstream the ancestry proofs logic in the future.

### CODE REVIEW

**Comment by serban300** [comment]
Converted to draft. Need to check some test failures

**Comment by serban300** [comment]
> Converted to draft. Need to check some test failures

Fixed

**Comment by acatangiu** [comment]
cc @lederstrumpf

File: src/ancestry_proof.rs [hunk 1]
```diff
@@ -136,109 +135,78 @@ impl<T: Clone + PartialEq, M: Merge<Item = T>> NodeMerkleProof<T, M> {
...
-                else if let Some(&(front_pos, ..)) = queue.front() {
-                    if height > 0 && is_descendant_pos(sib_pos, front_pos) {
-                        queue.push_back((pos, item, height));
-                        continue;
-                    } else {
-                        return Err(Error::CorruptedProof);
-                    }
                 } else {
+                    // let sibling_item = &proof_iter.next().ok_or(Error::CorruptedProof)?.1;
+                    // M::merge(sibling_item, &item)?
```
Comment (acatangiu) [inline src/ancestry_proof.rs]: remove leftover?
```suggestion
```

File: src/ancestry_proof.rs [hunk 2]
```diff
@@ -136,109 +135,78 @@ impl<T: Clone + PartialEq, M: Merge<Item = T>> NodeMerkleProof<T, M> {
...
-                } else if let Some(&(front_pos, ..)) = queue.front() {
-                    if height > 0 && is_descendant_pos(sib_pos, front_pos) {
-                        queue.push_back((pos, item, height));
-                        continue;
-                    } else {
-                        return Err(Error::CorruptedProof);
-                    }
                 } else {
+                    // let sibling_item = &proof_iter.next().ok_or(Error::CorruptedProof)?.1;
+                    // M::merge(&item, sibling_item)?
```
Comment (acatangiu) [inline src/ancestry_proof.rs]: same here?
```suggestion
```

File: src/ancestry_proof.rs [hunk 3]
```diff
@@ -321,3 +288,41 @@ fn calculate_root<
     let peaks_hashes = calculate_peaks_hashes::<_, M, _>(nodes, mmr_size, proof_iter)?;
     bagging_peaks_hashes::<_, M>(peaks_hashes)
 }
+
+pub fn expected_ancestry_proof_size(prev_mmr_size: u64, mmr_size: u64) -> usize {
```
Comment (acatangiu) [inline src/ancestry_proof.rs:290]: Will this be used outside of tests?

Do you plan to validate proofs in upper layers based on their expected size?

**Review by acatangiu — APPROVED** [review]
high-level looks good, would like an audit too for peace of mind

File: src/ancestry_proof.rs (hunk 3 above)
Comment (serban300) [reply src/ancestry_proof.rs:290]: Yes, I plan to call this from the `report_fork_voting()` extrinsic exposed in the beefy pallet, in order to check if the provided ancestry proof is optimal.

File: src/ancestry_proof.rs (hunk 1 above)
Comment (serban300) [reply src/ancestry_proof.rs]: This is the code used in `mmr.rs`. I just thought maybe it's good to keep it commented, just to see the difference between `mmr.rs` and `ancestry_proof.rs` very clear.

Comment (acatangiu) [reply src/ancestry_proof.rs]: would be worth a comment for it then specifying this

Comment (serban300) [reply src/ancestry_proof.rs]: Done

File: src/ancestry_proof.rs [hunk 4]
```diff
@@ -136,109 +135,82 @@ impl<T: Clone + PartialEq, M: Merge<Item = T>> NodeMerkleProof<T, M> {
...
-    let mut queue: VecDeque<_> = nodes
+    let mut queue: VecDeque<_> = VecDeque::new();
+    for value in nodes
         .into_iter()
         .map(|(pos, item)| (pos, item, pos_height_in_tree(pos)))
-        .collect();
-
-    let mut sibs_processed_from_back = Vec::new();
+    {
+        if !queue.insert_sorted_by(value, cmp_nodes) {
```
Comment (ordian) [inline src/ancestry_proof.rs]: previously, the q was sorted by the pos first and now it's (height, pos)
my understanding is that the `nodes` will be sorted by `pos`, so this algorithm might be running in $\mathcal{O}({nˆ2})$, is that correct?

File: src/ancestry_proof.rs [hunk 5]
```diff
@@ -321,3 +294,41 @@ fn calculate_root<
     let peaks_hashes = calculate_peaks_hashes::<_, M, _>(nodes, mmr_size, proof_iter)?;
     bagging_peaks_hashes::<_, M>(peaks_hashes)
 }
+
+pub fn expected_ancestry_proof_size(prev_mmr_size: u64, mmr_size: u64) -> usize {
```
Comment (ordian) [inline src/ancestry_proof.rs:290]: given the title of the PR, this seems like the main change in the PR. however, the rest seems like a totally unrelated rewriting of algorithm/refactoring. any reason to mix the two in one PR? or is it not a refactoring and actually this function depends on it?

also, what is the complexity of this function (seems non-trivial to guesstimate given nested loops)

**Review by ordian — COMMENTED** [review]
since i'm new to the code, left a couple of questions mainly about algorithmic complexity. 
still need some time to get an understanding of why this change preserves validity

File: src/ancestry_proof.rs (hunk 4 above)
Comment (serban300) [reply src/ancestry_proof.rs]: In the worst case scenario we want to prove all the nodes. Let's say the number of nodes = n. The complexity of inserting in an ordered queue is O(log(n)). So I think the total complexity would be O(n * log(n)).

But if we want to prove just one node, we need to start from it and go up until we reach the root. Let's say the number of leaves = l. The height of the MMR is log(l). The number of nodes in the proof is at most the height of the MMR (we need at most one node per level) so the queue length is at most log(l). So the complexity should be O(log(l) * log(l))

I hope I'm not missing anything.

File: src/ancestry_proof.rs (hunk 5 above)
Comment (serban300) [reply src/ancestry_proof.rs:290]: > given the title of the PR, this seems like the main change in the PR. 

Yes, exactly.

> however, the rest seems like a totally unrelated rewriting of algorithm/refactoring. any reason to mix the two in one PR? or is it not a refactoring and actually this function depends on it?

Yes, it's just a refactoring. No particular reason. Just did it for readability. I tried to follow the previous implementation to understand how to compute the expected proof size and it was just very hard. So I thought it might be helpful to refactor that as well. Also if we're doing a security audit for this PR, maybe it would be good to cover both changes. Also we'll need to upstream ancestry proofs at some point, and this change would help there, because it makes the code more similar to the upstream.

> also, what is the complexity of this function (seems non-trivial to guesstimate given nested loops)

Worst case scenario, for each peak we need to walk from a node inside it to the peak height. So should be something like O (num_peaks * log(num_leafs))

File: src/ancestry_proof.rs (hunk 4 above)
Comment (ordian) [reply src/ancestry_proof.rs]: > The complexity of inserting in an ordered queue is O(log(n)).

This is not true. Insertion at a random position in a VecDeque is $\mathcal{O}(n)$, because if you insert in the middle, you need to [shift](https://doc.rust-lang.org/1.80.1/src/alloc/collections/vec_deque/mod.rs.html#1868) every element on the right. If you want to guarantee logarithmic complexity, I'd suggest using a BTreeSet instead of a VecDeque.
But if the number in the proof is always bounded to a small number, that's probably fine as is. Just want to make sure we are not introducing a DoS vector in the worst case.

Comment (serban300) [reply src/ancestry_proof.rs]: Good point, you're right. It's O(n), because it's a vec-based structure. I missed that.

For our use case (ancestry proofs), yes the number of nodes in the proof should be small, but we should probably make this scalable in general. I will change it to a BTreeSet or a LinkedList.

Comment (ordian) [reply src/ancestry_proof.rs]: LinkedList would still have $\mathcal{O}(n)$ for finding the insertion position, so the right choice is BTreeSet or BinaryHeap 👍🏽

Comment (serban300) [reply src/ancestry_proof.rs]: > LinkedList would still have O ( n ) for finding the insertion position,

If it's sorted I think it should have O(log(n)). But yes, `BTreeSet` would have O(1) amortized. It's only that for small numbers it would be worse than a sorted structure. And usually the number of nodes should be small.
BinaryHeap is also a good option.

Comment (serban300) [reply src/ancestry_proof.rs]: Or sorry, in a linked list it's not that easy to take advantage of the fact that it's sorted. You're right.

Comment (serban300) [reply src/ancestry_proof.rs]: Yes, BTreeSet/BTreeMap since to be the best option. Done.

File: src/util.rs [hunk 6]
```diff
@@ -35,17 +35,18 @@ impl<T> MMRStoreWriteOps<T> for &MemStore<T> {
...
-            Err(pos) => self.insert(pos, value),
+impl<K: Ord, V: PartialEq> BTreeMapExt<K, V> for BTreeMap<K, V> {
+    fn checked_insert(&mut self, key: K, value: V) -> bool {
+        let old_value = self.get(&key);
+        if old_value != None && old_value != Some(&value) {
+            return false;
         }
+
+        self.insert(key, value);
+        true
```
Comment (ordian) [inline src/util.rs]: I'd suggest using entry API: https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.entry to avoid doing a lookup before insertion (or alternatively, clone the value and do a single insert)

**Review by ordian — APPROVED** [review]
Still don't fully grasp the details (probably best to ask @Lederstrumpf), but let's kick off the audit.

File: src/util.rs (hunk 6 above)
Comment (serban300) [reply src/util.rs]: Done

File: src/ancestry_proof.rs [hunk 7]
```diff
@@ -321,3 +286,41 @@ fn calculate_root<
...
+                if peak_idx <= peaks.len() {
+                    // Account for rhs bagging peaks
+                    expected_proof_size += 1;
+                }
+                break;
+            }
+        }
+    }
+
+    expected_proof_size
```
Comment (Lederstrumpf) [inline src/ancestry_proof.rs:325]: This works by checking the frequency of left-sibling-hood when iterating up from the last `prev_peak`.
In some sense, it's a myopic climber ascending a dark mountain and feeling out its shape from relative location :P
The mountain peaks here are the co-path items for proving said last `prev_peak`.
We can also determine the exact shape of the mountain from just knowing its width (i.e. the number of leaves beneath it) though, and skip the climbing step.

I've opened up #7 to show the changes, but feel free to cherry pick the new algorithm over here instead.

Comment (serban300) [reply src/ancestry_proof.rs:325]: Thank you ! Sounds good ! I'll take a look.

**Comment by serban300** [comment]
Closing in favor of https://github.com/paritytech/merkle-mountain-range/pull/7
